### PR TITLE
inductor: add device and dtype check before doing cpu fx packed weight

### LINKED ISSUE #98979: [torch.compile] raises `dense_to_mkldnn expects float or bfloat16 tensor input`
### 🐛 Describe the bug

`torch.compile` raises `dense_to_mkldnn expects float or bfloat16 tensor input` after doing some optimization

```py
import torch
import torch.nn as nn

torch.manual_seed(420)

class Model(torch.nn.Module):

    def __init__(self):
        super(Model, self).__init__()
        self.linear = torch.nn.Linear(2, 2).to(torch.float64)

    def forward(self, x):
        x = x.permute(1, 0)
        x = self.linear(x)
        x = x.permute(1, 0)
        return x
input_tensor = torch.rand(2, 2).to(torch.float64)

func = Model().to('cpu')

print(func(input_tensor))
# tensor([[-1.0019, -0.4457],
#         [ 0.1512, -0.4111]], dtype=torch.float64, grad_fn=<PermuteBackward0>)

with torch.no_grad():
    func.train(False)
    jit_func = torch.compile(func)
    print(jit_func(input_tensor))
    # torch._dynamo.exc.BackendCompilerFailed: backend='debug_wrapper' raised:
    # RuntimeError: dense_to_mkldnn expects float or bfloat16 tensor input
```

If we don't run it in `torch.no_grad` nor set `train` as `False`, it will succeed

```py
import torch
import torch.nn as nn

torch.manual_seed(420)

class Model(torch.nn.Module):

    def __init__(self):
        super(Model, self).__init__()
        self.linear = torch.nn.Linear(2, 2).to(torch.float64)

    def forward(self, x):
        x = x.permute(1, 0)
        x = self.linear(x)
        x = x.permute(1, 0)
        return x
input_tensor = torch.rand(2, 2).to(torch.float64)

func = Model().to('cpu')

print(func(input_tensor))
# tensor([[-1.0019, -0.4457],
#         [ 0.1512, -0.4111]], dtype=torch.float64, grad_fn=<PermuteBackward0>)

jit_func = torch.compile(func)
print(jit_func(input_tensor))
# tensor([[-1.0019, -0.4457],
#         [ 0.1512, -0.4111]], dtype=torch.float64,
#        grad_fn=<CompiledFunctionBackward>)
```

### Versions

```
Collecting environment information...
PyTorch version: 2.1.0.dev20230404+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: 14.0.0-1ubuntu1
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.9.16 (main, Mar  8 2023, 14:00:05)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.19.5-051905-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.5.119
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060
Nvidia driver version: 510.108.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          24
On-line CPU(s) list:             0-23
Vendor ID:                       GenuineIntel
Model name:                      12th Gen Intel(R) Core(TM) i9-12900K
CPU family:                      6
Model:                           151
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       1
Stepping:                        2
CPU max MHz:                     6700.0000
CPU min MHz:                     800.0000
BogoMIPS:                        6374.40
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault cat_l2 invpcid_single cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       640 KiB (16 instances)
L1i cache:                       768 KiB (16 instances)
L2 cache:                        14 MiB (10 instances)
L3 cache:                        30 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-23
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] numpy==1.24.1
[pip3] pytorch-triton==2.1.0+46672772b4
[pip3] torch==2.1.0.dev20230404+cu118
[pip3] torchaudio==2.1.0.dev20230404+cu118
[pip3] torchvision==0.16.0.dev20230404+cu118
[conda] numpy                     1.24.1                   pypi_0    pypi
[conda] pytorch-triton            2.1.0+46672772b4          pypi_0    pypi
[conda] torch                     2.1.0.dev20230404+cu118          pypi_0    pypi
[conda] torchaudio                2.1.0.dev20230404+cu118          pypi_0    pypi
[conda] torchvision               0.16.0.dev20230404+cu118          pypi_0    pypi
```

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen

### PR DESCRIPTION
**Pull request by XiaobingSuper**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99028

fix https://github.com/pytorch/pytorch/issues/96406 and https://github.com/pytorch/pytorch/issues/98979

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -1485,6 +1485,28 @@ def fn(a, b, c):
...
 
+    # https://github.com/pytorch/pytorch/issues/98979
+    def test_linear_float64(self):
+        mod = torch.nn.Sequential(torch.nn.Linear(8, 16).to(torch.float64)).eval()
+        with torch.no_grad():
+            self.common(mod, (torch.randn(2, 8).to(torch.float64),))
+
+    # https://github.com/pytorch/pytorch/issues/96406
+    @requires_cuda()
+    def test_linear_mixed_device(self):
```
Comment (jgong5) [inline test/inductor/test_torchinductor.py:1496]: Maybe move to `test_cuda_repro.py`?

**Review by jgong5 — APPROVED** [review]

File: test/inductor/test_torchinductor.py [hunk 2]
```diff
@@ -1485,6 +1485,13 @@ def fn(a, b, c):
             ),
         )
 
+    # https://github.com/pytorch/pytorch/issues/98979
+    @unittest.skipIf(HAS_CUDA, "cuda failed for float64 linear")
```
Comment (XiaobingSuper) [inline test/inductor/test_torchinductor.py:1489]: I skip this case for cuda path which it is always failed.  @jansel @desertfire @ngimel

File: test/inductor/test_torchinductor.py (hunk 1 above)
Comment (XiaobingSuper) [reply test/inductor/test_torchinductor.py]: yes, moved.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**